src/main/java/build/buildfarm/common/redis/RedisClient.java

```diff
@@ -109,8 +155,57 @@ public <T> T blockingCall(JedisInterruptibleContext<T> withJedis)
     return result;
   }
 
-  @SuppressWarnings("ConstantConditions")
   public <T> T call(JedisContext<T> withJedis) throws IOException {
+    return callImpl(withJedis);
+  }
+
+  private <T> T callImpl(JedisContext<T> withJedis) throws IOException {
+    // Capture all redis problems at the client level.
+    // Try to re-establish the client and log all issues.
+    // This will block the overall thread until redis can be connected to.
+    // It may be a useful strategy for gaining stability on a poorly performing network,
+    // or a redis cluster that goes down.  For example, if you need to perform redis maintance on a
+    // production cluster and do not want to affect build clients.
+    for (int i = 0; i < reconnectClientAttempts; ++i) {
+      try {
+        return defaultCall(withJedis);
+      } catch (Exception redisException) {
+        // Record redis failure.
+        redisErrorCounter.inc();
+        log.log(
+            Level.SEVERE,
+            String.format(
+                "Failure in RedisClient::callImpl (attempt %d of %d)",
+                i + 1, reconnectClientAttempts));
+        log.log(Level.SEVERE, redisException.toString());
+
+        // Wait before restablishing the client and trying again.
+        try {
+          Thread.sleep(reconnectClientWaitDurationMs);
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+        }
+        rebuildJedisCluser();
+      }
+    }
+
+    // If all attempts to re-establish the redis client has failed,
+    // perform the redis command and allow the exception to be thrown.
+    return defaultCall(withJedis);
+  }
+
+  private void rebuildJedisCluser() {
+    try {
+      log.log(Level.SEVERE, "Rebuilding redis client");
```

Comment:
Is this redundant with the log in thr upstream call? E.g. the caller already logged when it catches?

Reply:
agree, removed